pytket/extensions/cutensornet/mps/mps.py

```diff
@@ -62,33 +65,12 @@ class CuTensorNetHandle:
     def __init__(self, device_id: Optional[int] = None):
         self.handle = cutn.create()
         self._is_destroyed = False
-        dev = cp.cuda.Device(device_id)
-        self.device_id = int(dev)
-
-        if cp.cuda.runtime.runtimeGetVersion() < 11020:
-            raise RuntimeError("Requires CUDA 11.2+.")
-        if not dev.attributes["MemoryPoolsSupported"]:
-            raise RuntimeError("Device does not support CUDA Memory pools")
-
-        # Avoid shrinking the pool
-        mempool = cp.cuda.runtime.deviceGetDefaultMemPool(dev.id)
-        if int(cp.__version__.split(".")[0]) >= 10:
-            # this API is exposed since CuPy v10
-            cp.cuda.runtime.memPoolSetAttribute(
-                mempool,
-                cp.cuda.runtime.cudaMemPoolAttrReleaseThreshold,
-                0xFFFFFFFFFFFFFFFF,  # = UINT64_MAX
-            )
 
-        # A device memory handler lets CuTensorNet manage its own GPU memory
-        def malloc(size, stream):  # type: ignore
-            return cp.cuda.runtime.mallocAsync(size, stream)
-
-        def free(ptr, size, stream):  # type: ignore
-            cp.cuda.runtime.freeAsync(ptr, stream)
+        # Make sure CuPy uses the specified device
+        cp.cuda.Device(device_id).use()
 
-        memhandle = (malloc, free, "memory_handler")
-        cutn.set_device_mem_handler(self.handle, memhandle)
+        dev = cp.cuda.Device()

```

Comment:
In line 70 `cp.cuda.Device(device_id).use()` I've set cupy to use `device_id`, so that any call to `cp` stuff will use it by default. In particular, `cp.cuda.Device()` does as well, so you may think that `dev == device_id` everytime. However, the sneaky bit is that `device_id` may be `None`. In that case, `cp.cuda.Device(device_id).use()` is simple saying "use the default device" and `dev = cp.cuda.Device()` is returning the id of said default device.

> Do you think this class is still usable for other purposes such as intrinsic MPI support for the backend?

Yeah, as usable as it was before. We'd just need to keep multiple instances of `CuTensorNetHandle`, each on a different device. The non-trivialy part would still be to refactor the `MPS` class itself so that we can keep track of which tensors are on which device, so that we know which of these `CuTensorNetHandle`s should be used when applying operations on them. And, of course, what to do with message passing when acting between tensors in different devices (possibly, just sending one of the tensors to the other device, updating and then sending back).

Reply:
Yeah, I am thinking about this more and more often - sounds like we really need it. Yes, I think we can use a common stencil-like approach as the first attempt. I will hope to start looking into this soonish.